### DISCUSSION TITLE
Improvements to Symbols Tree in React components

### DISCUSSION THREAD

**Discussion by ghost** [zed-industries/zed#24001] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

The symbols tree in React pages and components is a little lackluster, it would be great to have a more detailed outline of the component I am working on, for example see the Zed symbols tree compared to JetBrains Fleet's for the same file:

<img width="254" alt="Screenshot 2024-11-28 at 14 37 36" src="https://github.com/user-attachments/assets/37b06d2f-3742-4145-9c5b-0f149124612e">
<img width="337" alt="Screenshot 2024-11-28 at 14 37 47" src="https://github.com/user-attachments/assets/5463fe5d-0334-4071-939b-4c70a5f8e611">

### Environment

Zed: v0.163.2 (Zed)
OS: macOS 15.1.0
Memory: 18 GiB
Architecture: aarch64

**Comment by maxbrunsfeld** [comment]
Can you share the source file for this example?

**Comment by ghost** [comment]
Hey @maxbrunsfeld sorry about the delay, here is a subset of the source file as it's quite a large file:
``` ts
/**
 * Builder component that renders a page based on the given `content` array.
 *
 * The `content` array is expected to be an array of objects with a `_type`
 * property that determines which component to render.
 */
const Builder = async ({ content }: BuilderProps) => {
  const siteSettings = await getSiteSettings();

  return (
    <>
      {content?.map(async (block, index) => {
        switch (block._type) {
          // Hero Carousel
          case "heroCarousel":
            return (
              <HeroCarousel
                key={index}
                slides={block.slides || []}
                isRolex={isRolex}
              />
            );

          // Featured Items
          case "featuredItems":
            return <FeaturedItems key={index} items={block.items || []} />;

          // Image Call to Action
          case "imageCta":
            return (
              <ImageCta
                key={index}
                imageSrc={
                  urlFor(block.image || { _type: "image" })?.url() as string
                }
                imageAlt={block.title as string}
                title={block.title as string}
                description={block.description as string}
                linkHref={block.href as string}
                linkText={block.text as string}
              />
            );

          default:
            return null;
        }
      })}
    </>
  )
};

export default Builder;

```

**Comment by maxbrunsfeld** [comment]
Ok, so the idea is that we'd include an outline entry for every JSX element expression?

**Comment by ghost** [comment]
Pretty much with obvious folding at the root of each nested tree of elements. 

Just makes navigating a react file a little more convenient when you can jump to symbols.

**Comment by yasinkavakli** [comment]
Can I bump this?
Just take this file https://raw.githubusercontent.com/vercel/nextjs-postgres-nextauth-tailwindcss-template/refs/heads/main/app/(dashboard)/page.tsx from a Next.js template and you'll see this in Zed:
<img width="296" alt="image" src="https://github.com/user-attachments/assets/babca66a-e062-4452-bc47-179a4fa0e1e3" />
and this in VSCode: 
<img width="320" alt="image" src="https://github.com/user-attachments/assets/fcdf5c15-3e61-4605-b2b6-a5e4a65b24ef" />

I've fully switched to Zed but this is really something that annoys me and I miss from VSCode.
